File: packages/python/goldenanalysis/goldenanalysis/analyzers/match_rates.py
"""``match.rates`` — match metrics from a GoldenMatch result's artifacts.

Reads ``scored_pairs`` / ``match_stats`` (+ optional ``recall_certificate`` and
``match_threshold``) from ``AnalyzerInput.artifacts``. Degrades: emits the metrics
its present artifacts support and omits the rest.
"""

from __future__ import annotations

from typing import Any

from goldenanalysis.core import aggregate as agg
from goldenanalysis.models import (
    AnalysisTable,
    AnalyzerInfo,
    AnalyzerInput,
    AnalyzerResult,
    Metric,
)

_PRODUCES = [
    "match.pair_count",
    "match.match_rate",
    "match.threshold",
    "match.recall_estimate",
    "match.recall_safe_bound",
    "match.mean_pair_score",
]


def _cert_values(cert: Any) -> tuple[float | None, float | None]:
    """Normalize a certificate (dict or RecallEstimate/RecallCertificate) to
    ``(estimate, safe_bound)``. Either may be None."""
    if cert is None:
        return None, None
    if isinstance(cert, dict):
        return cert.get("estimate", cert.get("recall")), cert.get("safe_bound", cert.get("recall_lower"))
    # dataclass: RecallEstimate has .recall; RecallCertificate adds .recall_lower
    return getattr(cert, "recall", None), getattr(cert, "recall_lower", None)


class MatchRatesAnalyzer:
    """Pair counts, match rate, recall (from a certificate), score distribution."""

    info = AnalyzerInfo(name="match.rates", consumes=["scored_pairs", "match_stats"], produces=_PRODUCES)

    def run(self, inp: AnalyzerInput) -> AnalyzerResult:
        art = inp.artifacts
        scored_pairs = art.get("scored_pairs", [])
        stats = art.get("match_stats", {}) or {}

        metrics: list[Metric] = [
            Metric(key="match.pair_count", value=len(scored_pairs), unit="pairs", direction="neutral"),
        ]
        if "match_rate" in stats:
            metrics.append(
                Metric(key="match.match_rate", value=stats["match_rate"], unit="ratio", direction="neutral")
            )
        if art.get("match_threshold") is not None:
            metrics.append(
                Metric(key="match.threshold", value=art["match_threshold"], unit="score", direction="neutral")
            )

        estimate, safe_bound = _cert_values(art.get("recall_certificate"))
        if estimate is not None:
            metrics.append(
                Metric(key="match.recall_estimate", value=estimate, unit="ratio", direction="higher_better")
            )
        if safe_bound is not None:
            metrics.append(
                Metric(
                    key="match.recall_safe_bound", value=safe_bound, unit="ratio", direction="higher_better"
                )
            )

        tables: list[AnalysisTable] = []
        if scored_pairs:
            scores = [float(s) for *_, s in scored_pairs]
            mean_score = sum(scores) / len(scores)
            metrics.append(
                Metric(key="match.mean_pair_score", value=mean_score, unit="score", direction="neutral")
            )
            hist = agg.histogram(scores, bins=10)
            tables.append(
                AnalysisTable(
                    name="score_histogram",
                    columns=["bin_left", "count"],
                    rows=[[edge, count] for edge, count in hist],
                )
            )

        return AnalyzerResult(metrics=metrics, tables=tables)

File: packages/typescript/goldenanalysis/src/core/analyzers/matchRates.ts
/**
 * `match.rates` — match metrics from a GoldenMatch result's artifacts.
 *
 * Reads `scored_pairs` / `match_stats` (+ optional `recall_certificate` and
 * `match_threshold`) from `AnalyzerInput.artifacts` — the SAME snake_case keys the
 * Python sibling reads, so a serialized `PipeResult.artifacts` feeds this identically.
 * Degrades: emits the metrics its present artifacts support and omits the rest.
 * Parity with `packages/python/goldenanalysis/goldenanalysis/analyzers/match_rates.py`.
 */

import { histogram } from "../aggregate.js";
import type {
  AnalysisTable,
  Analyzer,
  AnalyzerInfo,
  AnalyzerInput,
  AnalyzerResult,
  Metric,
} from "../types.js";

const PRODUCES = [
  "match.pair_count",
  "match.match_rate",
  "match.threshold",
  "match.recall_estimate",
  "match.recall_safe_bound",
  "match.mean_pair_score",
];

function asNumber(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

/** Normalize a certificate (dict-ish) to [estimate, safeBound]; either may be null. */
function certValues(cert: unknown): [number | null, number | null] {
  if (cert === null || cert === undefined || typeof cert !== "object") return [null, null];
  const c = cert as Record<string, unknown>;
  return [asNumber(c["estimate"] ?? c["recall"]), asNumber(c["safe_bound"] ?? c["recall_lower"])];
}

export class MatchRatesAnalyzer implements Analyzer {
  readonly info: AnalyzerInfo = {
    name: "match.rates",
    consumes: ["scored_pairs", "match_stats"],
    produces: PRODUCES,
  };

  run(input: AnalyzerInput): AnalyzerResult {
    const art = input.artifacts;
    const scoredPairs = Array.isArray(art["scored_pairs"]) ? (art["scored_pairs"] as unknown[]) : [];
    const stats = (art["match_stats"] as Record<string, unknown> | undefined) ?? {};

    const metrics: Metric[] = [
      { key: "match.pair_count", value: scoredPairs.length, unit: "pairs", direction: "neutral" },
    ];
    if ("match_rate" in stats) {
      metrics.push({ key: "match.match_rate", value: Number(stats["match_rate"]), unit: "ratio", direction: "neutral" });
    }
    const threshold = art["match_threshold"];
    if (threshold !== null && threshold !== undefined) {
      metrics.push({ key: "match.threshold", value: Number(threshold), unit: "score", direction: "neutral" });
    }

    const [estimate, safeBound] = certValues(art["recall_certificate"]);
    if (estimate !== null) {
      metrics.push({ key: "match.recall_estimate", value: estimate, unit: "ratio", direction: "higher_better" });
    }
    if (safeBound !== null) {
      metrics.push({ key: "match.recall_safe_bound", value: safeBound, unit: "ratio", direction: "higher_better" });
    }

    const tables: AnalysisTable[] = [];
    if (scoredPairs.length > 0) {
      // Each pair is (...ids, score) — the score is the last element.
      const scores = scoredPairs.map((p) => {
        const arr = p as unknown[];
        return Number(arr[arr.length - 1]);
      });
      const meanScore = scores.reduce((a, b) => a + b, 0) / scores.length;
      metrics.push({ key: "match.mean_pair_score", value: meanScore, unit: "score", direction: "neutral" });
      const hist = histogram(scores, 10);
      tables.push({
        name: "score_histogram",
        columns: ["bin_left", "count"],
        rows: hist.map(([edge, count]) => [edge, count]),
      });
    }

    return { metrics, tables };
  }
}
